result

ADIGrimm opened the pull request.

ADIGrimm committed:
result

mateacademy-ai-mentor reviewed (approved):
Great job! 🎉 Your implementation meets the task requirements and constraints effectively. The code is well-structured, and all necessary components are correctly configured. Just make sure to include Swagger documentation if it's required elsewhere in your project. Keep up the excellent work! 🚀

✨ Please rate this review in the Platform's notification center. I will be grateful for your feedback! ✨

Elena-Bruyako commented on src/main/java/mate/academy/rickandmorty/config/AppConfig.java:
```diff
@@ -0,0 +1,15 @@
+public class AppConfig {
+
+    @Bean
+    public RestTemplate restTemplate() {
```
you actually don't need to create bean for RestTemplate by this way, so you can remove this config

Elena-Bruyako commented on src/main/java/mate/academy/rickandmorty/mapper/CharacterMapper.java:
```diff
@@ -0,0 +1,34 @@
+
+@Component
+public class CharacterMapper {
+    public CharacterDto toDto(Character character) {
```
don't need to implement mapping by yourself, use MapStruct

Elena-Bruyako commented on src/main/java/mate/academy/rickandmorty/model/Character.java:
```diff
@@ -0,0 +1,23 @@
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    private String internalId;
```
```suggestion
    @Column(nullable = false, unique = true)
    private String internalId;
```

Elena-Bruyako commented on src/main/java/mate/academy/rickandmorty/service/impl/CharacterServiceImpl.java:
```diff
@@ -0,0 +1,70 @@
+
+    @Override
+    public List<ApiDto> fetchAllCharacters() {
+        String url = "https://rickandmortyapi.com/api/character?page=";
```
it's better to use class-level variable for url
Also, you can use `@Value` and add url to app properties

Elena-Bruyako commented on src/main/java/mate/academy/rickandmorty/dto/ApiDto.java:
```diff
@@ -0,0 +1,30 @@
+import lombok.Data;
+
+@Data
+public class ApiDto {
```
For this task such Dto will be enough

Example, 
```
public record CharacterResponseDto(
        Map<String, Object> info,
        List<CharacterExternalDto> results) {
}

public record CharacterExternalDto(
        Long id,
        String name,
        String status,
        String gender) {
}

public record CharacterDto(
        Long id,
        Long externalId,
        String name,
        String status,
        String gender
) {}
```

ADIGrimm committed:
fixed

ahoienko commented on pom.xml:
```diff
@@ -15,12 +15,20 @@
+			<artifactId>mysql-connector-j</artifactId>
+			<scope>runtime</scope>
+		</dependency>
+
```
```suggestion
```

ahoienko commented on pom.xml:
```diff
@@ -32,6 +40,17 @@
+			<groupId>org.springframework</groupId>
+			<artifactId>spring-web</artifactId>
+		</dependency>
+
```
```suggestion
```

ahoienko commented on src/main/java/mate/academy/rickandmorty/mapper/CharacterMapper.java:
```diff
@@ -0,0 +1,20 @@
+    Character toEntity(CharacterExternalDto characterResponseDto);
+
+    default String generateInternalId(CharacterExternalDto dto) {
+        return dto.getId() != null ? String.valueOf(dto.getId()) : "default-id";
```
```suggestion
        return dto.getId() != null ? String.valueOf(dto.getId()) : DEFAULT_ID;
```

ahoienko replied:
Not resolved

ahoienko commented on src/main/java/mate/academy/rickandmorty/service/impl/CharacterServiceImpl.java:
```diff
@@ -0,0 +1,69 @@
+@Service
+@RequiredArgsConstructor
+public class CharacterServiceImpl implements CharacterService {
+    public static final String URL = "https://rickandmortyapi.com/api/character?page=";
+    public static final Long CHARACTER_MAX = 826L;
```
```suggestion
    private static final String URL = "https://rickandmortyapi.com/api/character?page=";
    private static final Long CHARACTER_MAX = 826L;
```

ahoienko replied:
Not resolved

ahoienko commented on src/main/java/mate/academy/rickandmorty/service/impl/CharacterServiceImpl.java:
```diff
@@ -0,0 +1,69 @@
+
+    @Override
+    public CharacterDto getRandomCharacter() {
+        Random random = new Random();
```
What is the reason to create this object each time method is called?

ahoienko replied:
Not resolved

ahoienko commented on src/main/java/mate/academy/rickandmorty/service/impl/CharacterServiceImpl.java:
```diff
@@ -0,0 +1,69 @@
+
+    @Override
+    public List<CharacterExternalDto> fetchAllCharacters() {
+        List<CharacterExternalDto> allCharacters = new ArrayList<>();
+        int page = 1;
+        CharacterResponseDto response;
+        do {
+            ResponseEntity<CharacterResponseDto> apiResponse = restTemplate
+                    .getForEntity(URL + page, CharacterResponseDto.class);
+            response = apiResponse.getBody();
+            if (response != null && response.getResults() != null) {
+                allCharacters.addAll(response.getResults());
+            }
+            String nextPageUrl = (String) response.getInfo().get("next");
+            page++;
+        } while (response != null && response.getInfo().get("next") != null);
+        return allCharacters;
```
move string literals to constants

ahoienko replied:
Not resolved

ahoienko commented on src/main/java/mate/academy/rickandmorty/service/impl/CharacterServiceImpl.java:
```diff
@@ -0,0 +1,69 @@
+
+    @Override
+    public List<CharacterExternalDto> fetchAllCharacters() {
+        List<CharacterExternalDto> allCharacters = new ArrayList<>();
+        int page = 1;
+        CharacterResponseDto response;
+        do {
+            ResponseEntity<CharacterResponseDto> apiResponse = restTemplate
+                    .getForEntity(URL + page, CharacterResponseDto.class);
+            response = apiResponse.getBody();
+            if (response != null && response.getResults() != null) {
+                allCharacters.addAll(response.getResults());
+            }
+            String nextPageUrl = (String) response.getInfo().get("next");
+            page++;
+        } while (response != null && response.getInfo().get("next") != null);
+        return allCharacters;
```
Add some line breaks

ahoienko replied:
Not resolved

ahoienko commented on src/main/java/mate/academy/rickandmorty/service/impl/CharacterServiceImpl.java:
```diff
@@ -0,0 +1,69 @@
+        return allCharacters;
+    }
+
+    @PostConstruct
```
PostConstruct annotation is used to indicate 'init' method, the porpose of init method is to configure something in bean, code that you wrote here has nothing to deal with bean configuration. If you want to make some operation on program start please check ApplicationReadyEvent or use CommandLineRunner

ahoienko replied:
Not resolved

ahoienko commented on src/main/resources/db/changelog/db.changelog-master.yaml:
```diff
@@ -0,0 +1,3 @@
+databaseChangeLog:
+  - include:
+      file: db/changelog/changes/01-create-characters-table.yaml
```
```suggestion
      file: db/changelog/changes/01-create-characters-table.yaml
      
```

ahoienko replied:
Not resolved

ahoienko reviewed (changes requested):
Nice, left some comments

ADIGrimm committed:
fixed

added/removed lines in CharacterServiceImpl, db.changelog-master, pom
added constants in CharacterServiceImpl, CharacterMapper

ahoienko reviewed (changes requested):
Please resolve comments before pushing new solution

ADIGrimm commented:
Sorry, I forgot to push
